src/providers/spatialite/qgsspatialitefeatureiterator.cpp

```diff
@@ -584,7 +584,7 @@ QVariant QgsSpatiaLiteFeatureIterator::getFeatureAttribute( sqlite3_stmt *stmt,
   }
 
   // assuming NULL
-  return QVariant( type );
+  return QVariant( );
```

Comment:
Is that required? Won't that create an invalid QVariant instead of a Null QVariant?

I know we often use them interchangeably, but I was always hoping that in the long run we'll get a clean coherent usage for them. That would open up some nice possibilities for "we know the attribute is NULL" vs. "the attribute was not requested / not yet initialized (still needs default values)" and similar.

Reply:
Not that it makes much sense (all passing here on 5.11.2):
```c++
  QVERIFY( QVariant().isNull() );
  QVERIFY( !QVariant().isValid() );
  QCOMPARE(QVariant().toString(), QString( "" ) );
  QVERIFY( QVariant(QVariant::LongLong).isNull() );
  QVERIFY( QVariant(QVariant::LongLong).isValid() );
  QCOMPARE(QVariant(QVariant::LongLong).toString(), QString( "0" ) );
```

Btw, this was the main source of the bug because there are a lot of checks for the QVariant type == field.type()  and a conversion to the field type passing through the string representation if the type does not match.

This one-liner change costed me an afternoon (mostly spent on writing tests).

Reply:
> Btw, this was the main source of the bug because there are a lot of checks for the QVariant type == field.type() and a conversion to the field type passing through the string representation if the type does not match.

Isn't this the real problem? There should rather be an `isNull()` check before these. NULL QVariants are produced in many places throughout our codebase and should be accepted as such.

Reply:
I'll have a second look, is this the semantics?
- isNull() when the value has been set to NULL
- !isValid() when we the value has not been set

Reply:
@m-kuhn ok, I've found another solution that respects the null semantic, with all test cases in place it was much easier now!

Reply:
Good job @elpaso !
Thanks for the tests and the fix.